opm/autodiff/NewtonSolver.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright 2015 SINTEF ICT, Applied Mathematics.
+  Copyright 2015 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_NEWTONSOLVER_HEADER_INCLUDED
+#define OPM_NEWTONSOLVER_HEADER_INCLUDED
+
+#include <opm/autodiff/AutoDiffBlock.hpp>
+#include <opm/core/utility/parameters/ParameterGroup.hpp>
+
+namespace Opm {
+
+
+    /// A Newton solver class suitable for general fully-implicit models.
+    template <class PhysicalModel>
+    class NewtonSolver
+    {
+    public:
+        // ---------  Types and enums  ---------
+        typedef AutoDiffBlock<double> ADB;
```

Comment:
(for newbies, the name `ADB` and `ADB::V` are not helpful at all. it's probably better to bite the bullet and write `AutoDiffBlock` and `ValueVector` (or whatever 'V' stands for). If you want to replace this, you can use a pretty simple sed script or ping me.)